MoreLinq/TrySingle.cs

```diff
@@ -0,0 +1,171 @@
+        /// <param name="zero">
+        /// The value that is passed as the first argument to
+        /// <paramref name="resultSelector" /> if the sequence has zero
+        /// elements.</param>
+        /// <param name="one">
+        /// The value that is passed as the first argument to
+        /// <paramref name="resultSelector" /> if the sequence has a
+        /// single element only.</param>
+        /// <param name="many">
+        /// The value that is passed as the first argument to
+        /// <paramref name="resultSelector" /> if the sequence has two or
+        /// more elements.</param>
+        /// <param name="resultSelector">
+        /// A function that is provided with the cardinality, and if the
+        /// sequence has just one element, the value of that element. Then
+        /// transforms the result to an instance of TResult.</param>
+        /// <typeparam name="T">
+        /// The type of the elements of the sequence.</typeparam>
+        /// <typeparam name="TCardinality">
+        /// The type that expresses cardinality.</typeparam>
+        /// <typeparam name="TResult">
+        /// The result type of the <paramref name="resultSelector"/> function.
+        /// </typeparam>
+        /// <returns>
+        /// The value returned by the <paramref name="resultSelector"/>.
+        /// </returns>
+        /// <remarks>
+        /// This operator uses immediate execution, but never consumes more
+        /// than two elements from the sequence. When the source sequence is an
+        /// <see cref="IList{T}"/> or <see cref="ICollection{T}"/> then the
+        /// implementation optimizes by checking the number of elements in
+        /// the underlying sequence.</remarks>
+        public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
+            TCardinality zero, TCardinality one, TCardinality many,
+            Func<TCardinality, T, TResult> resultSelector)
+        {
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            switch (source)
+            {
+                case IReadOnlyCollection<T> readOnlyCollection:
+                    return TrySingleForReadOnlyCollection(readOnlyCollection);
+                case ICollection<T> collection:
+                    return TrySingleForCollection(collection);
+                default:
+                    return TrySingleForEnumerable(source);
+            }
+
+            TResult TrySingleForReadOnlyCollection(IReadOnlyCollection<T> theCollection)
+            {
+                switch (theCollection.Count)
+                {
+                    case 0:
+                        return resultSelector(zero, default);
+                    case 1:
+                        return resultSelector(one,
+                            theCollection is IReadOnlyList<T> theList ? theList[0]
```

Comment:
I know it's odd to say this, but just because a type implements `IReadOnlyCollection<>` doesn't mean it will implement `IReadOnlyList<>` and not `IList<>`. One should cater for both.